### DISCUSSION TITLE
[Python] Altering Fixed parameter, amici.setFixedParameters()

### DISCUSSION THREAD

**Discussion by Podde1** [AMICI-dev/AMICI#1442] [Q&A] [answered]
Hi! 

I have recently started using the Python implementation of amici (I have previously used the MATLAB version) and I have now found myself a bit stuck. 

I am working with an SBML model, importing and simulation the model works fine. The issue I am having is that I am not able to change the value of a fixed parameter in my model. The `amici.setFixedParameters()` function (and the others with similar functionality, -`ById()`, -`ByName()`)  runs and generate no error. However, the value of my parameter is not changed when I simulate the model. 
Since `amici.setParameters()` and `amici.setInitialStates()` works as intended and my fixed parameter is found by `amici.getFixedParameterIds()` I assume that the value of my fixed-parameter can't be overwritten. 

As previously stated I work with an SBML model. The fixed parameter is defined as follows (I am using Tellurium to create the SBML model): 
`k_stim := piecewise(10, time <5, 0);`
Meaning that it is a time-sensitive parameter that changes value when the time-statement is true. To experiment I tried stating the parameter as `k_stim := 0;`. This had no effect and `amici.setFixedParameters()` runs without errors but do not change the value of the fixed parameter. 

Am I interpreting something wrong? Are there any other ways to alter my fixed parameter in my amici model object? 

Thanks in advance! 
//Henrik

**Comment by paulstapor** [comment]
Hi Henrik,

the problem is, as you guessed, that you're parameter is actually time-dependent.
Die you implement the time-dependence via an assignment rule?

The implementation of _parameters_ with assignment rules in AMICI is maybe not yet fully functional, I far as I know. But it's on our list. However, you could use a _species_ with an assignment rule to circumvent the problem for the moment. This should hopefully do the trick.
Alternatively, we're working on an event implementation right now. Depending on how this will look like in detail, you might also implement your model using an event at the trigger timepoint, altering a time-constant species...

Best,
Paul

**Reply by FFroehlich** [reply]
Hi Henrik,

> Hi!
> 
> I have recently started using the Python implementation of amici (I have previously used the MATLAB version) and I have now found myself a bit stuck.

Great, happy to help!

> 
> I am working with an SBML model, importing and simulation the model works fine. The issue I am having is that I am not able to change the value of a fixed parameter in my model. The `amici.setFixedParameters()` function (and the others with similar functionality, -`ById()`, -`ByName()`) runs and generate no error. However, the value of my parameter is not changed when I simulate the model.
> Since `amici.setParameters()` and `amici.setInitialStates()` works as intended and my fixed parameter is found by `amici.getFixedParameterIds()` I assume that the value of my fixed-parameter can't be overwritten.
> 

That's surprising. These functions should usually all work and throw and error if they don't. I assume what is happening is that something went wrong during model import and you did not end up with the structure you intended to have.

> As previously stated I work with an SBML model. The fixed parameter is defined as follows (I am using Tellurium to create the SBML model):
> `k_stim := piecewise(10, time <5, 0);`
> Meaning that it is a time-sensitive parameter that changes value when the time-statement is true. To experiment I tried stating the parameter as `k_stim := 0;`. This had no effect and `amici.setFixedParameters()` runs without errors but do not change the value of the fixed parameter.

Do I understand correctly that you are trying to set the value of `k_stim` using something like `model.setFixedParameterByName("k_stim", 0.0)`? AMICI does not support setting complete expressions to new values at runtime, since this would require recompilation of sensitivity equations, so you cannot set the value of `k_stim` if you previously defined it as `k_stim := piecewise(10, time <5, 0)`. It looks like AMICI does not implement the required sanity checks since this should have produced an error when compiling the model (what probably ended up happening is that AMICI created a FixedParameter that, as you noticed, has no influence on the model whatsoever).

We will implement a fix such that an error is thrown during model import.

I hope this helps,
Best,
Fabian

> 
> Am I interpreting something wrong? Are there any other ways to alter my fixed parameter in my amici model object?
> 
> Thanks in advance!
> //Henrik

**Reply by Podde1** [reply]
Hi @paulstapor , thanks for the replay. 

> the problem is, as you guessed, that you're parameter is actually time-dependent.
> Die you implement the time-dependence via an assignment rule?

Yes I did, the `:=` annotation declares an assignment rule. 

> However, you could use a species with an assignment rule to circumvent the problem for the moment. This should hopefully do the trick.

Hm, that might do the trick. I got to try if Tellurium allows for an assignment rule of constant species. 

> Alternatively, we're working on an event implementation right now. 

That's great! The whole reason I am messing around with the assignment rule on the `k_stim` parameter is that I was not allowed to import the model with an event that triggered the state, since `sbml2amici()` did not have any support for this.

**Reply by Podde1** [reply]
Hi @FFroehlich , thanks for the reply. 

> >I am working with an SBML model, importing and simulation the model works fine. The issue I am having is that I am not able to change the value of a fixed parameter in my model. The amici.setFixedParameters() function (and the others with similar functionality, -ById(), -ByName()) runs and generate no error. However, the value of my parameter is not changed when I simulate the model.
Since amici.setParameters() and amici.setInitialStates() works as intended and my fixed parameter is found by amici.getFixedParameterIds() I assume that the value of my fixed-parameter can't be overwritten.

>That's surprising. These functions should usually all work and throw and error if they don't. I assume what is happening is that something went wrong during model import and you did not end up with the structure you intended to have.

As far as I can tell my model is imported correctly, it produces the same results as when I simulate the model in an SBML-software. Additionally, I can find all spieces, initial conditions, parameters and fixed parameters using the `get...()` functions on the `amici.model` object. 

> Do I understand correctly that you are trying to set the value of k_stim using something like model.setFixedParameterByName("k_stim", 0.0)? AMICI does not support setting complete expressions to new values at runtime, since this would require recompilation of sensitivity equations, so you cannot set the value of k_stim if you previously defined it as k_stim := piecewise(10, time <5, 0). It looks like AMICI does not implement the required sanity checks since this should have produced an error when compiling the model (what probably ended up happening is that AMICI created a FixedParameter that, as you noticed, has no influence on the model whatsoever).

Well, yes what I have been trying is to set `model.setFixedParameterByName("k_stim", 0.0)` and that it would require recompilation of sensitivity equations  makes sense now that you mention it, 

As I replied to Paul above, the `k_stim` parameter is just a proxy solution since events are not implemented in the sbml-importer. But until events are implemented I need a workaround, and ultimately I would like to declare `model.setFixedParameterByName("k_stim", numpy.piecewise(x, condlist, funclist))`, to change the piecewise-function. However, this would not be possible, I assume, from your previously reply? (Also this would require me to access the time variable from the model object, which I have not been able to find). 
The reason why I want to change the k_stim parameter comes from how my model is constructed, I will try to declare it shortly below:

My model starts at steady-state and is using a constant spiece to apply a stimulation effect on my system. I would like to be able to change how and when this stimulation effect acts on the system, to not create "x" number of the same model with different stimulation effects.   
I don't really mind how I make this work, k_stim was just the first solution I found to achieve this stimulation effect. If events are the only solution to this behaviour, it would be neat to be able to change the events from the model object.  

Best regards 
// Henrik

**Reply by Podde1** [reply]
Hi again @paulstapor, 

I have tested to move my assignment rule to a spice instead of a parameter, which works with regard to importing the model. However, the spiece is considered an expression in the amici model object and as far as I am aware there is no way to alter the value of an expression i.e. there are no `amici.model.setExpressionValues()` function.

**Comment by Podde1** [comment, marked as answer]
With amici version, 0.11.14, SBMLimport of events is now supported and can be used instead of assignment rules. However, there is currently no way to alter either events or assignment rules in the model object. 
Therefore, to be able to do so different SBMl-files are required currently.
